### PR TITLE
refactor: move hermetic Python runtime setup into macro

### PR DESCRIPTION
**Pull request by rickeylev**
This stems from conversation in #2216. Moving the logic into a loading-phase macro makes it
easier to understand how a repo with a python-build-standalone extracted archive has
its toolchain defined. This also makes it easier to compare with the local runtime repo
setup to try and find commonality in the future.

Along the way:
* Remove some Bazel 5 compatibility code: coverage_tool is now always passed to py_runtime.
  (Bazel 5 support was dropped earlier this year and we stopped testing with it; it's
  likely already broken with Bazel 5).
* Add `semver.to_dict`, which makes it easier to pass into format() functions.

### CODE REVIEW

File: python/private/hermetic_runtime_repo_setup.bzl [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+            format.
+        python_bin: {type}`str` The path to the Python binary within the
+            repositoroy.
+        coverage_tool: {type}`str` optional target to the coverage tool to
+            use.
+    """
+    _ = name  # @unused
+    python_version_info = python_version.split(".")
+    major = python_version_info[0]
+    minor = python_version_info[1]
```
Comment (aignas) [inline python/private/hermetic_runtime_repo_setup.bzl:53]: Nit, using the `semver` function from #2218 would be nice here.

File: python/private/hermetic_runtime_repo_setup.bzl [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Setup a python-build-standalone based toolchain."""
+
+load("@rules_cc//cc:defs.bzl", "cc_import", "cc_library")
+load("@rules_python//python:py_runtime.bzl", "py_runtime")
+load("@rules_python//python:py_runtime_pair.bzl", "py_runtime_pair")
+load("@rules_python//python/cc:py_cc_toolchain.bzl", "py_cc_toolchain")
+load("@rules_python//python/private:py_exec_tools_toolchain.bzl", "py_exec_tools_toolchain")
```
Comment (aignas) [inline python/private/hermetic_runtime_repo_setup.bzl:20]: I think the loads here could be simplified:
```suggestion
load("//python:py_runtime.bzl", "py_runtime")
load("//python:py_runtime_pair.bzl", "py_runtime_pair")
load("//python/cc:py_cc_toolchain.bzl", "py_cc_toolchain")
load("//python/private:py_exec_tools_toolchain.bzl", "py_exec_tools_toolchain")
```

File: python/private/hermetic_runtime_repo_setup.bzl [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    # Used to only download coverage toolchain when the coverage is collected by
+    # bazel.
+    native.config_setting(
+        name = "coverage_enabled",
+        values = {"collect_code_coverage": "true"},
+        visibility = ["//visibility:private"],
+    )
+    py_runtime_kwargs = {}
+    if coverage_tool:
+        # Earlier versions of Bazel don't support this attribute.
```
Comment (aignas) [inline python/private/hermetic_runtime_repo_setup.bzl:127]: I think it is supported from `6.0` onwards that is the lowest supported version, so we can drop the `if` here.

File: python/private/hermetic_runtime_repo_setup.bzl [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        files = [":files"],
+        interpreter = python_bin,
+        interpreter_version_info = {
+            "major": str(major),
+            "micro": str(python_version_info[2]),
+            "minor": str(minor),
+        },
+        python_version = "PY3",
+        implementation_name = "cpython",
+        pyc_tag = "cpython-{}{}".format(major, minor),
```
Comment (aignas) [inline python/private/hermetic_runtime_repo_setup.bzl:141]: I am so used to this being something like `cp{major}{minor}` in the wheels. It's interesting that this is different, should we have a small comment here linking to some docs?

File: python/private/hermetic_runtime_repo_setup.bzl [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    py_cc_toolchain(
+        name = "py_cc_toolchain",
+        headers = ":python_headers",
+        libs = ":libpython",
+        python_version = python_version,
+    )
+
+    py_exec_tools_toolchain(
+        name = "py_exec_tools_toolchain",
+        precompiler = "@rules_python//tools/precompiler:precompiler",
```
Comment (aignas) [inline python/private/hermetic_runtime_repo_setup.bzl:160]: nit: not using `@rules_python`
```suggestion
        precompiler = Label("//tools/precompiler:precompiler"),
```

**Review by aignas — APPROVED** [review]
Only minor things, so LGTM!

File: python/private/hermetic_runtime_repo_setup.bzl (hunk 3 above)
Comment (rickeylev) [reply python/private/hermetic_runtime_repo_setup.bzl]: Done

File: python/private/hermetic_runtime_repo_setup.bzl (hunk 2 above)
Comment (rickeylev) [reply python/private/hermetic_runtime_repo_setup.bzl]: Done

File: python/private/hermetic_runtime_repo_setup.bzl (hunk 4 above)
Comment (rickeylev) [reply python/private/hermetic_runtime_repo_setup.bzl]: Linked to the pep

File: python/private/hermetic_runtime_repo_setup.bzl (hunk 5 above)
Comment (rickeylev) [reply python/private/hermetic_runtime_repo_setup.bzl]: Done

**Pull request merged**